### ISSUE TITLE
Users & Authentication -- `Roles` bugs & enhancements

### ISSUE THREAD

**Issue by slickwarren** [rancher/dashboard#2675]
the below were found using 2.5-head (d1083bc63e4ddde70ffadd09bf0133ae5862b5fe) with 2.5-dev UI as a restricted admin with `manage roles` global permissions enabled:
### Functionality Issues

- [ ] upon editing as yaml, user is brought to a page that doesn't have tabs for `global` , `cluster`, etc. 
<img width="1792" alt="Screen Shot 2021-04-12 at 11 31 44 AM" src="https://user-images.githubusercontent.com/16691014/114443883-14fbcf00-9b83-11eb-9004-020c70a7e06d.png">

<img width="1792" alt="Screen Shot 2021-04-12 at 11 31 52 AM" src="https://user-images.githubusercontent.com/16691014/114443788-f5fd3d00-9b82-11eb-8940-9d59d6b55bf3.png">
-- clicking the `Roles` Page brings user back to normal page

- [ ] ~~multiple defaults allowed~~ for a given role 'type' (i.e. cluster) first created role will be used as the true default when creating user/assigning roles (updated based on first comment below)
<img width="812" alt="Screen Shot 2021-04-12 at 11 11 19 AM" src="https://user-images.githubusercontent.com/16691014/114444114-59876a80-9b83-11eb-9a16-765d757c5ec0.png">

- [ ] when using the `inherit from` option and have entered a valid role to inherit from, `grant resources` has required fields, but are ignored
<img width="856" alt="Screen Shot 2021-04-12 at 10 44 56 AM" src="https://user-images.githubusercontent.com/16691014/114444212-70c65800-9b83-11eb-926c-2b2f01a8cc7d.png">
-- note: users are already able to create an empty role by removing the first required resource and saving the role, which has the same outcome as this being ignored

- [ ] restricted admin is logged out when attempting to edit an existing built-in role in cluster or project/namespace role type
-- restricted admin does not have `manage roles` permissions, but should just not have access instead of being logged out on attempt to edit 
-- restricted admin was still able to successfully create and edit new (non-built-in) roles, as well as delete built-in roles for cluster/project/namespace role types
-- adding the `manage roles` role does allow user to edit built-in roles successfully
-- there are no `resources` to choose from project/namespace roles... entering your own, even with `manage roles` logs out user

- [ ] project/namespaces, unable to view/edit created role(s)
Repro by:
* create project/namespace role (there aren't any resources to choose from, but the custom create for `resources` works 
* save the role
* view new role

```
vue.runtime.esm.js:1888 TypeError: Cannot read property 'rules' of undefined
    at f.createRules (RoleDetailEdit.vue:255)
    at RoleDetailEdit.vue:286
    at Array.forEach (<anonymous>)
    at f.createInheritedRules (RoleDetailEdit.vue:280)
    at f.inheritedRules (RoleDetailEdit.vue:205)
    at _n.get (vue.runtime.esm.js:4479)
    at _n.evaluate (vue.runtime.esm.js:4584)
    at f.inheritedRules (vue.runtime.esm.js:4836)
    at f.r (RoleDetailEdit.vue?f278:1)
    at f.t._render (vue.runtime.esm.js:3548)
```
<img width="990" alt="Screen Shot 2021-04-12 at 1 26 41 PM" src="https://user-images.githubusercontent.com/16691014/114457363-ca824e80-9b92-11eb-863a-15e82e61e95d.png">

### Visual/Display Issues
- [ ] `resources` I think should be labeled `resource` since user can only select 1 resource per resource grant. This field should be marked as required, as an error is displayed in the UI if it is not filled  HOWEVER there are some resource grants from built-in roles that do have blank resources... so I might be misunderstanding the error. . 
<img width="264" alt="Screen Shot 2021-04-12 at 10 41 18 AM" src="https://user-images.githubusercontent.com/16691014/114444456-c995f080-9b83-11eb-9ed5-9d395808322d.png">

- [ ] as per comments in https://github.com/rancher/dashboard/pull/2656#issuecomment-815914890, custom roles should also display in the dropdown once created
<img width="267" alt="Screen Shot 2021-04-12 at 10 39 30 AM" src="https://user-images.githubusercontent.com/16691014/114444695-0530ba80-9b84-11eb-99f3-5d358740b670.png">

- [ ] when viewing a created resource, radio buttons are no longer filled (makes it slightly difficult to see which was selected on first glance) 
<img width="961" alt="Screen Shot 2021-04-12 at 10 35 32 AM" src="https://user-images.githubusercontent.com/16691014/114444890-3b6e3a00-9b84-11eb-8a7b-e3bc14c3b0a3.png">

- [ ] `Locked` does not have a default selected and does not send a status for `Locked` on the backend (should default to no/false)
<img width="695" alt="Screen Shot 2021-04-12 at 10 42 35 AM" src="https://user-images.githubusercontent.com/16691014/114444405-b420c680-9b83-11eb-971c-a6332a16de8a.png">
-- seems to save OK, probably defaults to false on k8s side

**Comment by vincent99** [comment]
>  multiple defaults allowed for a given role 'type' (i.e. cluster) (first one selected will be used as the true default when creating user/assigning roles)

Multiple default roles are allowed/normal.  A new user (global) or the user who creates a project/cluster should get ALL of them. (In Ember the column describes this slightly better as New User Default, Cluster Creator Default, Project Creator Default)

**Comment by richard-cox** [comment]
> ### Functionality Issues

> * [x]  upon editing as yaml, user is brought to a page that doesn't have tabs for `global` , `cluster`, etc.

Fixed by #2677

> * [x]  ~multiple defaults allowed~ for a given role 'type' (i.e. cluster) first created role will be used as the true default when creating user/assigning roles (updated based on first comment below)

As per Vince's comment, there can be multiple defaults for all types. I've updated the title of the 'Default' column for each type to better represent it's meaning and also match the label used when configuring. PR #2677

> * [x]  when using the `inherit from` option and have entered a valid role to inherit from, `grant resources` has required fields, but are ignored

The API is pretty forgiving and roles can be created with no rules or inherited details. To represent this I've removed the required indicator for the role. PR #2677

> * [x]  restricted admin is logged out when attempting to edit an existing built-in role in cluster or project/namespace role type

The only way I could reproduce this was at the project level when specifying junk resources that the user didn't not have permissions to see. Cause by the dashboard interpreting any 401 error as session related and logs the user out. We now no longer do this and the error is shown. PR #2677

> * [ ]  project/namespaces, unable to view/edit created role(s)

I was unable to reproduce this, could you specify the global user roles and form values?

> ### Visual/Display Issues

> * [x]  `resources` I think should be labeled `resource` since user can only select 1 resource per resource grant. This field should be marked as required, as an error is displayed in the UI if it is not filled  HOWEVER there are some resource grants from built-in roles that do have blank resources... so I might be misunderstanding the error. .

Fixed by #2677

> * [ ]  as per comments in [#2656 (comment)](https://github.com/rancher/dashboard/pull/2656#issuecomment-815914890), custom roles should also display in the dropdown once created

This should be a separate ticket. The taggable setting is applied to both the 'verb' select control (where multiple values can be selected) and the 'resource' select control (where only a single value can be selected, and entered text filters drop down).

> * [x]  when viewing a created resource, radio buttons are no longer filled (makes it slightly difficult to see which was selected on first glance)

Fixed by #2677. The default radio button will not show it's normal colour with an opacity, rather than a muted grey one. Have also fixed the cursor for the disabled radio button blob.

> * [x]  `Locked` does not have a default selected and does not send a status for `Locked` on the backend (should default to no/false)

Fixed by #2677

**Comment by anupama2501** [comment]
Verified the following issues on v2.5-31f00c15e511b6d588bd0958e4873f818baa558c-head

1. - [X]  upon editing as yaml, user is brought to a page that doesn't have tabs for global , cluster, etc
From cluster explorer --> Users and authentication --> Users --> create a user with restricted admin permission enabled and manage roles enabled under the roles
Log in as restricted admin and cluster explorer --> Users and authentication --> Roles --> global create a role --> Cluster create a role --> Project/namespace create a role.
Edit the roles as yaml and when clicked on save, users are re-directed to the roles page with global,clusters and Project/namespaces.

2. - [X]  multiple defaults allowed for a given role 'type' (i.e. cluster) first created role will be used as the true default when creating user/assigning roles (updated based on first comment below)
For each role i.e., Global, cluster, project - multiple defaults are allowed and for each role, the header specifies the type of defaults
<img width="300" alt="Screen Shot 2021-04-19 at 10 09 16 AM" src="https://user-images.githubusercontent.com/38144301/115276191-95787d80-a0f7-11eb-91fb-865543a05464.png">

3. - [] when using the inherit from option and have entered a valid role to inherit from, grant resources has required fields, but are ignored

- [X] > i. From projects, create a role and click on inherit from and select any of the resources and grant resources is not required option and the save goes through without any issue


4.  restricted admin is logged out when attempting to edit an existing built-in role in cluster or project/namespace role type
- [X]-- restricted admin does not have manage roles permissions, but should just not have access instead of being logged out on attempt to edit
On editing, restricted admin is not logged out and displayed a warning.
-- restricted admin was still able to successfully create and edit new (non-built-in) roles, as well as delete built-in roles for cluster/project/namespace role types.
Creation and deletion of new roles on the restricted admin went through fine without any issues.
-- adding the manage roles role does allow user to edit built-in roles successfully
Enabled mange roles on the user s restricted admin and the user was able to edit the roles. 


Visual/Display Issues:
1. Resources under the role --> create roles is displaying resource for global, cluster and Project
2.  Radio buttons are visible while viewing the created roles.
3.  Locked default is set to No

**Comment by anupama2501** [comment]
Re-opening the issue as the following are still seen:
3. when using the inherit from option and have entered a valid role to inherit from, grant resources has required fields, but are ignored
From projects, create a role and click on inherit from and select any of the resources and grant resources is not required option and the save goes through without any issue
 ii. Now click on the grant resources and in the resources no matching resources were found. 

4. Enabled mange roles on the users restricted admin and the user was able to edit the roles. 
-- there are no resources to choose from project/namespace roles... entering your own, even with manage roles logs out user
this is still not showing and resources for the new roles.

5. project/namespaces, unable to view/edit created role(s)
Repro by:
create project/namespace role (there aren't any resources to choose from, but the custom create for resources works
save the role
view new role

Custom create option is not shown in the resources option.
If we login as restricted admin, the resources are not shown for the users even the custom option.

**Comment by richard-cox** [comment]
> Re-opening the issue as the following are still seen:
> 3. when using the inherit from option and have entered a valid role to inherit from, grant resources has required fields, but are ignored
> From projects, create a role and click on inherit from and select any of the resources and grant resources is not required option and the save goes through without any issue
> ii. Now click on the grant resources and in the resources no matching resources were found.

I'm not sure of the instructions here, as an admin I can create the role as expected
1. Navigate to Users & Authentication --> Roles --> Project/Namespaces Tab
1. Click `Create a Project/Namespaces Role`
2. Click 'Inherit From` tab, Add Resource, Select a Resource
3. Click `Grant Resources` tab and no required fields are indicated
4. Add a Name for the role at the top
5. Click `Create` 
6. Roles Page shown on `Project/Namespaces` tab, with new role shown when sorted by Age

 
> 1. Enabled mange roles on the users restricted admin and the user was able to edit the roles.
>    -- there are no resources to choose from project/namespace roles... entering your own, even with manage roles logs out user
>    this is still not showing and resources for the new roles.
If the user does not have visibility of namespaced resources that are not `cattle.io` types the resource list in Grant Resources will be empty, I think this is to be expected.

@codyrancher This uses the code as was (below), though it looks like this is an update from Ember. Is this still ok?
```
const GLOBAL_GROUP = '.cattle.io';
...

  allResources() {
    return this.$getters['all'](SCHEMA).filter(r => r.attributes?.kind);
  },

  namespaceResources() {
    return this.allResources.filter(r => r.attributes.namespaced && !r.attributes.group.includes(GLOBAL_GROUP));
  },
```

> 2. project/namespaces, unable to view/edit created role(s)
>    Repro by:
>    create project/namespace role (there aren't any resources to choose from, but the custom create for resources works
>    save the role
>    view new role
> 
> Custom create option is not shown in the resources option.
> If we login as restricted admin, the resources are not shown for the users even the custom option.

As per above, restricted admins will only see specific resources. For user supplied `Resource` values, unlike ember, there is no `(custom)` resource to select. To supply a custom resource the user needs to type their value and hit return. As a restricted admin, with the manage role global role, I've managed to successfully created a `Project/Namespaes` role with a custom Resource.

**Comment by codyrancher** [comment]
@richard-cox I think it's fine.

**Comment by richard-cox** [comment]
Ok, to summaries.. after chatting with Anupama the only remaining issue here is that a `restricted-admin` with the `manage-roles` global role cannot see options in a rule's  `Resources` field when creating a `Project/Namespaces` role.

![image](https://user-images.githubusercontent.com/18697775/116236468-03a3ec80-a757-11eb-8b6c-f97056432493.png)

Previously in ember it looks like these options were a predefined [fixed set](https://github.com/rancher/ui/blob/6463676c1f718a135567e37e09cc4724f107b3a2/lib/shared/addon/utils/constants.js#L398)

We now calculate the options via
```
export const CATTLE_API_GROUP = '.cattle.io';
...
allResources() {
  return this.$getters['all'](SCHEMA).filter(r => r.attributes?.kind);
},
namespaceResources() {
  return this.allResources.filter(r => r.attributes.namespaced && !r.attributes.group.includes(CATTLE_API_GROUP));
},
```
So if there are no namespaced resources (that are not Rancher's) visible then there will be no resources to select. The user is still able to enter their own custom resource by simply typing it in.

While testing, I did manage to get resources to show by adding the `restricted-user` to a project.

**Comment by richard-cox** [comment]
Discussed with Vince. We SHOULD hardcode Cluster and Namespace role rule resource lists.. we don't know which cluster they'll be applied to so can't got out and fetch acceptable resource type suggestions. #2855

**Comment by sowmyav27** [comment]
Reopening for a PR in 2.5.8

**Comment by anupama2501** [comment]
Re-opening the issue for PR in 2.5.8 branch as the above PR is merged in master

**Comment by richard-cox** [comment]
@sowmyav27 Please accept my apologies, fix went into master and not 2.5.8 as well. This has now been done rancher/dashboard#2875

**Comment by anupama2501** [comment]
On v2.5.8-rc16 verified the following scenarios:

> when using the inherit from option and have entered a valid role to inherit from, grant resources has required fields, but are ignored
From projects, create a role and click on inherit from and select any of the resources and grant resources is not required option and the save goes through without any issue
ii. Now click on the grant resources and in the resources no matching resources were found.

- Logged in as restricted admin, with manage roles enabled. Created a role in cluster explorer --> Roles --> projects/namespaces
- from the inherit from, select any of the options available
- From grant resources, selected the verbs "delete" and the resources user - roles were created successfully

> Enabled mange roles on the users restricted admin and the user was able to edit the roles.
-- there are no resources to choose from project/namespace roles... entering your own, even with manage roles logs out user
this is still not showing and resources for the new roles.

- Resources are shown for the new roles and created roles with the resources option


> project/namespaces, unable to view/edit created role(s)

- custom roles can be added and for the custom roles, all the resources are made available.
